File: src/main/java/me/mucloud/plugin/MK/EasyShop/core/Shop.java
package me.mucloud.plugin.MK.EasyShop.core;

import me.mucloud.plugin.MK.EasyShop.api.IShop;
import me.mucloud.plugin.MK.EasyShop.api.Viewable;
import me.mucloud.plugin.MK.EasyShop.internal.Messages;
import org.bukkit.Material;
import org.bukkit.entity.Player;
import org.bukkit.inventory.Inventory;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.ItemMeta;;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;


public class Shop<T extends Product> implements IShop<T>, Viewable{

    private boolean Enabled;
    private String ShopName;
    private ShopPool.ShopType ShopType;
    private Material ICON;
    private List<String> ShopDetail;
    private String AccessPermission;
    private ShopRefreshTask RefreshTask;
    private final List<T> ProductionList;

    Shop(ShopPool.ShopType shopType, String shopName, Material icon, List<String> shopDetail, @Nullable String accessPermission, int refreshInterval){
        ShopType = shopType;
        ShopName = shopName;
        ICON = icon;
        ShopDetail = shopDetail;
        AccessPermission = accessPermission;
        ProductionList = new ArrayList<>();
        if(refreshInterval != -1){
            RefreshTask = new ShopRefreshTask(this);
        }

        Enabled = true;
    }

    @Override public ShopPool.ShopType getShopType(){
        return ShopType;
    }

    @Override public boolean isEnabled(){
        return Enabled;
    }

    @Override public void close() {
        Enabled = false;
    }

    @Override public void open() {
        Enabled = true;

    }

    @Override public void setShopName(String shopName){
        ShopName = shopName;
    }

    @Override public String getShopName(){
        return ShopName;
    }

    @Override public Inventory toInv(Player p) {
        //todo
        return null;
    }

    @Override public ItemStack toIcon(){
        ItemStack is = new ItemStack(ICON, 1);
        ItemMeta im = is.getItemMeta();
        im.setDisplayName(ShopName);
        im.setLore(Arrays.asList(
                Messages.SHOP_REMAINING_PREFIX,
                getRemaining()
        ));
        is.setItemMeta(im);
        return is;
    }

    @Override public void setPermission(String permission){
        AccessPermission = permission;
    }

    @Override public boolean checkPermission(Player p){
        return AccessPermission == null || p.hasPermission(AccessPermission);
    }

    @Override public void setIcon(Material icon) {
        ICON = icon;
    }

    @Override public void refresh() {
        ProductionList.forEach(Product::refresh);
    }

    public String getRemaining(){
        //todo
        return null;
    }

    @Override public int getRefreshInterval(){
        //todo
        return 0;
    }

    @Override public void setRefreshTime(int interval){
        if(RefreshTask != null){
            RefreshTask.refreshInterval(interval);
        }
    }

    @Override public int addProduction(T production){
        for(Product p : ProductionList){
            if(p.equals(production)){
                return 1;
            }
        }
        ProductionList.add(production);
        return 0;
    }

    @Override public int removeProduction(T production){
        int index = -1;
        for(Product p : ProductionList){
            if(p.equals(production)){
                index = ProductionList.indexOf(p);
            }
        }
        if(index != -1){
            ProductionList.remove(index);
            return 0;
        }
        return 1;
    }

    @Override public List<T> toList(){
        return ProductionList;
    }

    @Override public Product getProduction(int index){
        return ProductionList.get(index);
    }

    @Override public boolean equals(String shopName){
        return ShopName.equals(shopName);
    }

    @Override public boolean equals(Shop<?> s) {
        return ShopName.equals(s.getShopName()) && ICON == s.ICON;
    }

}

File: CORE/src/main/kotlin/me/mucloud/mcplugin/MK/EasyShop/core/Shop.kt
package me.mucloud.mcplugin.MK.EasyShop.core

import me.mucloud.mcplugin.api.MK.EasyShop.Product
import me.mucloud.mcplugin.api.MK.EasyShop.Shop
import org.bukkit.Bukkit
import org.bukkit.Material
import org.bukkit.entity.Player
import org.bukkit.inventory.Inventory

class Shop(
    internal val id: Int,
    internal val icon: Material,
    internal var name: String,
    internal var description: String,
    internal var type: ShopType,
    internal var products: MutableList<Product> = mutableListOf()
): Shop {

    private var inv: Inventory? = null

    init {
        inv = Bukkit.createInventory(null, 5, name).also{ i ->
            products.forEach{ p ->
                i.addItem(p.toIcon())
            }
        }
    }

    override fun getID(): Int = id

    override fun addProduct(product: Product): Boolean {
        if(products.contains(product)){
            return true
        }else{
            products.add(product)
            return true
        }
    }

    override fun getShopName(): String = name

    override fun getShopDescription(): String = description

    override fun getOriginViewport(): Inventory? = inv

    override fun removeProduct(): Boolean {
        return true
    }

    override fun getProduct(id: Int): Product? {
        products.forEach{
            if(it.getID() == id){
                return it
            }
        }
        return null
    }

    override fun deal(player: Player, product: Product): Boolean {
        return true
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as me.mucloud.mcplugin.MK.EasyShop.core.Shop

        if (id != other.id) return false
        if (icon != other.icon) return false
        if (name != other.name) return false
        if (description != other.description) return false
        if (type != other.type) return false

        return true
    }

    override fun hashCode(): Int {
        var result = id
        result = 31 * result + icon.hashCode()
        result = 31 * result + name.hashCode()
        result = 31 * result + description.hashCode()
        result = 31 * result + type.hashCode()
        return result
    }


}